### ISSUE TITLE
Adaptive loss reweighting for PINNs

### ISSUE THREAD

**Issue by zoemcc** [SciML/NeuralPDE.jl#276]
I ran across this talk/paper recently that adaptively reweights the different components in the PINN loss (pde loss, bc loss, ic loss, etc) according to their recent gradient statistics so that they're all approximately the same scale range.  It uses a reweighting scheme similar to the mechanism within ADAM that rescales the components of the gradient during gradient descent.  This should help the optimization scheme make sure that all of the components of the PDE specified are given equal importance. This will help avoid situations where the optimization gets stuck in the local minimum of a uniformly flat solution that technically satisfies the PDE loss but ignores the BC or IC loss because the real solution lies far away on the loss surface.  The PDE loss is usually much higher magnitude than the BC or IC losses due to its derivative nature and so it's hard to escape from that minimum.

It's a pretty easy mechanism to implement, mathematically.  The main questions I have are related to the interface for this and not stepping on anyone's toes.

I think we want to pass it in as a kwarg option of the PINN:

```julia
NeuralPDE.PhysicsInformedNN(chain, strategy; adaptive_loss=AdaptiveLoss(10))
```

with an option struct:

```julia
mutable struct AdaptiveLoss
  reweight_every::Int64
  i::Int64
  loss_weights::Vector{Float32} # we seem to be hard coding Float32's often, this could be a parameterized type
  @add_kwonly function AdaptiveLoss(reweight_every, i=0)
    new(convert(Int64, reweight_every), convert(Int64, i), Float32[])
  end
end
```  

There are two main difficulties that I can see:

1.  The internal mechanism of the optimization function needs to know how many optimization iterations have occurred.  This is not necessarily the same as the number of times the optimization function has been called.  We could have the user increment some counter internal to the PINN through a callback, but this is hard to use and unwieldy.  Does an OptimizationFunction have some other mechanism to keep track of iteration count that I'm not aware of?  We want to make sure that this loss reweighting only occurs in between optimization iterations or undefined behavior could occur since the loss would appear to change unpredictably from the point of view of the top level optimization.

2. The paper suggests a separate scaling factor for each type of loss: ic, bc, pde, etc.  Since we want our framework to eventually support incredibly general PDEs with many different BCs and ICs and other conditions, I think we want each individual bc equation to get its own loss scale.  Unfortunately, the way we generate loss functions bundles up all of the bc losses and pde losses into one function to call.  The rescaling that I need to do is very similar to the existing tau argument that we pass into loss function generation but that parameter is viewed as a constant and not a parameter.  I need this to be a parameter so that we can call `bc_loss_function(θ, \tau)` in loss_function and use the correct scaling factors.  Is that a problem for anyone?

References:
[talk](https://www.youtube.com/watch?v=OFUkRgCI6No&list=PL1e3Jic2_DwwJQ528agJYMEpA0oMaDSA9&index=8)
[paper](https://arxiv.org/abs/2001.04536v1)
[code](https://github.com/PredictiveIntelligenceLab/GradientPathologiesPINNs)

**Comment by ChrisRackauckas** [comment]
https://github.com/SciML/NeuralPDE.jl/issues/155 also describes a few methods.

**Comment by asdasdqwdqwfsdf** [comment]
Hi @ChrisRackauckas

is it possible that we can set the weights for the boundary terms using NeuralPDE?

**Comment by ChrisRackauckas** [comment]
Not yet. We're working on that.

**Comment by asdasdqwdqwfsdf** [comment]
> 
> 
> Not yet. We're working on that.

Hi @ChrisRackauckas,

can we now use "MiniMax" function in "NeuralPDE.jl"?

**Comment by ChrisRackauckas** [comment]
Check the WIP PR.
